File: python/realtime/diarization.py
from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from diarization import Diarizer

    from .session import Session

log = logging.getLogger("realtime.diarization")

async def run_diarization(
    session: "Session",
    diarizer: "Diarizer | None",
    item_id: str,
    audio: list[float],
    audio_end_ms: int,
) -> None:
    if diarizer is None:
        return

    sr = 16_000
    utt_start_ms = max(0, audio_end_ms - (len(audio) * 1000) // sr)
    t0 = time.monotonic()
    try:
        try:
            import numpy as np

            arr = np.asarray(audio, dtype=np.float32)
        except ImportError:
            arr = audio
        segments = diarizer.diarize_utterance(arr, utt_start_ms)
    except Exception as err:
        log.warning("diarization failed: %s (item_id=%s)", err, item_id)
        return
    elapsed_ms = int((time.monotonic() - t0) * 1000)

    if not segments:
        return

    sink = await session.event_sink()
    if sink is None:
        return
    payload: dict[str, Any] = {
        "type": "conversation.item.diarization",
        "item_id": item_id,
        "audio_end_ms": int(audio_end_ms),
        "elapsed_ms": elapsed_ms,
        "segments": [
            {
                "speaker": f"SPEAKER_{int(s.speaker):02d}",
                "start": s.t_start_ms / 1000.0,
                "end": s.t_end_ms / 1000.0,
                "confidence": float(s.confidence),
            }
            for s in segments
        ],
    }
    await sink.send_value(payload)

File: rust/src/realtime/diarization.rs
use std::sync::Arc;
use std::sync::OnceLock;

use tracing::warn;

use super::inspector::InspectorEvent;
use super::session::Session;

pub(super) fn realtime_enabled() -> bool {
    static ENABLED: OnceLock<bool> = OnceLock::new();
    *ENABLED.get_or_init(
        || match std::env::var(crate::defaults::diarization::REALTIME_ENV) {
            Ok(v) => matches!(
                v.trim().to_ascii_lowercase().as_str(),
                "1" | "true" | "yes" | "on"
            ),
            Err(_) => crate::defaults::diarization::REALTIME_ENABLED,
        },
    )
}

pub(super) async fn run_diarization(
    session: Arc<Session>,
    item_id: String,
    audio: Vec<f32>,
    audio_end_ms: u64,
) {
    let mut guard = session.diarizer.lock().await;
    let diarizer = match guard.as_mut() {
        Some(d) => d,
        None => return,
    };

    let utt_start_ms = audio_end_ms.saturating_sub((audio.len() as u64) * 1000 / 16_000);
    let t0 = std::time::Instant::now();
    let result = diarizer.diarize_utterance(&audio, utt_start_ms);
    drop(guard);
    let elapsed_ms = t0.elapsed().as_millis() as u64;

    let segments = match result {
        Ok(s) => s,
        Err(e) => {
            let reason = format!("{e}");
            warn!(error = %reason, %item_id, "diarization failed");
            session.inspector.emit(InspectorEvent::DiarizationEmitted {
                session_id: session.id.as_str().to_string(),
                item_id: item_id.clone(),
                audio_end_ms,
                num_segments: 0,
                num_speakers: 0,
                elapsed_ms,
                failed: true,
                reason: Some(reason),
            });
            return;
        }
    };

    let mut speakers: std::collections::BTreeSet<u32> = std::collections::BTreeSet::new();
    for s in &segments {
        speakers.insert(s.speaker);
    }
    let num_segments = segments.len() as u32;
    let num_speakers = speakers.len() as u32;

    if !segments.is_empty() {
        session
            .emit(super::wire::OutboundEvent::Diarization {
                item_id: item_id.clone().into(),
                audio_end_ms,
                elapsed_ms: Some(elapsed_ms),
                segments: segments
                    .iter()
                    .map(|s| {
                        serde_json::json!({
                            "speaker": format!("SPEAKER_{:02}", s.speaker),
                            "start": s.t_start_ms as f64 / 1000.0,
                            "end": s.t_end_ms as f64 / 1000.0,
                            "confidence": s.confidence,
                        })
                    })
                    .collect(),
            })
            .await;
    }

    session.inspector.emit(InspectorEvent::DiarizationEmitted {
        session_id: session.id.as_str().to_string(),
        item_id,
        audio_end_ms,
        num_segments,
        num_speakers,
        elapsed_ms,
        failed: false,
        reason: None,
    });
}
